research-prototype/Programs/Source/ipae2e.mpc

```diff
@@ -0,0 +1,154 @@
+
+_list = Matrix(numrows, 4, sint)
+
+local_params = {
+    '_list': _list,
+    'sint': sint,
+    'print_ln': print_ln,
+}
+
+for i in range(numrows):
+    for j in range(4):
+        cmd = f"_list[{i}][{j}] = sint.get_input_from(0)"
+        exec(cmd, {}, local_params)
+
+# Radix sort
+_list.sort((0,), n_bits=32)
+
+if verbose:
+    for i in range(numrows):
+        cmd = (f"print_ln('%s, %s, %s, %s', "
+               f"_list[{i}][0].reveal(), _list[{i}][1].reveal(),"
+               f"_list[{i}][2].reveal(), _list[{i}][3].reveal(),)")
+        exec(cmd, {}, local_params)
+
+helperbit = Array(numrows, sint)
+local_params['helperbit'] = helperbit
+
+for i in range(numrows-1):
+    cmd = f"helperbit[{i}] = (_list[{i}][0] == _list[{i}+1][0]).if_else(1, 0)"
+    exec(cmd, {}, local_params)
+
+
+helperbit_and_istrigger = Array(numrows, sint)
+local_params['helperbit_and_istrigger'] = helperbit_and_istrigger
+helperbit_xor_istrigger = Array(numrows, sint)
+local_params['helperbit_xor_istrigger'] = helperbit_xor_istrigger
+
+for i in range(numrows-1):
+    cmd = f"helperbit_and_istrigger[{i}] = helperbit[{i}] * _list[{i}][1]"
+    exec(cmd, {}, local_params)
+    cmd = f"helperbit_xor_istrigger[{i}] = (helperbit[{i}] + _list[{i}][1]) % 2"
+    exec(cmd, {}, local_params)
+
+if verbose:
+    print_ln('%s', helperbit.reveal())
+    print_ln('%s', helperbit_and_istrigger.reveal())
+    print_ln('%s', helperbit_xor_istrigger.reveal())
+
+
+stopbits = [Array(numrows, sint) for _ in range(numrows_power)]
+local_params['stopbits'] = stopbits
+
+for stopbit in stopbits:
+    local_params['stopbit'] = stopbit
+    for i in range(numrows):
+        cmd = f"stopbit[{i}] = 1"
+        exec(cmd, {}, local_params)
+
+credits = [Array(numrows, sint) for _ in range(numrows_power)]
+local_params['credits'] = credits
+
+for i, (stopbit, credit) in enumerate(zip(stopbits, credits)):
+    stepsize = 2 ** i
+    local_params['stopbit'] = stopbit
+    local_params['credit'] = credit
+    local_params['stepsize'] = stepsize
+
+    for current in range(numrows - stepsize):
+        successor = current + stepsize
+        # another optimization could be to multiply one time all the
+        # helperbit * istrigger bits since these are fixed.
+        if i == 0:
+            cmd = f"flag = helperbit_and_istrigger[{successor}];"
+            # credit is initialized to value, _list[i][2] = value[i]
+            cmd += f"credit[{current}] = _list[{current}][2] + flag * _list[{successor}][2];"
+            # * stopbit[successor]  stopbit[successor] here would always be 1 so can drop
+            cmd += f"stopbit[{current}] = flag"
+        else:
+            cmd = f"flag = stopbits[{i-1}][{current}] * helperbit_and_istrigger[{successor}];"
+            cmd += f"credit[{current}] = "
+            cmd += f"credits[{i-1}][{current}] + flag * credits[{i-1}][{successor}];"
+            cmd += f"stopbit[{current}] = flag * stopbits[{i-1}][{successor}]"
```

Comment:
+= being used to concatenate a bunch of strings, which sort of loosely correspond to lines of code... is just not awesome.

Reply:
fixed.